libmobilecoin/src/chacha20_rng.rs

```diff
@@ -0,0 +1,149 @@
+    ffi_boundary_with_error(out_error, || {
+        let bytes: [u8; 32] = bytes.as_slice_of_len(32)?.try_into().map_err(|_| {
+            LibMcError::InvalidInput("seed bytes length must be exactly 32 bytes".to_owned())
+        })?;
+        Ok(Mutex::new(McChaCha20Rng::from_seed(bytes)))
+    })
+}
+
+/// Returns the current word_pos of the ChaCha20Rng instance
+///
+/// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+/// * `out_word_pos` - pointer to buffer of 16 bytes where the current
+///   chacha20_rng wordpos will be returned
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_get_word_pos(
+    chacha20_rng: FfiMutPtr<Mutex<McChaCha20Rng>>,
+    out_word_pos: FfiMutPtr<McMutableBuffer>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> bool {
+    ffi_boundary_with_error(out_error, || {
+        let word_pos = chacha20_rng.lock()?.get_word_pos();
+        let out_word_pos = out_word_pos.into_mut().as_slice_mut_of_len(16)?;
+        out_word_pos.copy_from_slice(&word_pos.to_be_bytes());
+        Ok(())
+    })
+}
+
+/// Sets the current word_pos of the ChaCha20Rng instance
+///
+/// /// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+/// * `out_word_pos` - pointer to buffer of 128 bytes where the current
+///   chacha20_rng wordpos will be returned
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_set_word_pos(
+    chacha20_rng: FfiMutPtr<Mutex<McChaCha20Rng>>,
+    bytes: FfiRefPtr<McBuffer>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> bool {
+    ffi_boundary_with_error(out_error, || {
+        let bytes: [u8; 16] = bytes.as_slice_of_len(16)?.try_into().map_err(|_| {
+            LibMcError::InvalidInput(
+                "bytes length must be exactly 16 bytes for word_pos".to_owned(),
+            )
+        })?;
+        chacha20_rng
+            .lock()?
+            .set_word_pos(u128::from_be_bytes(bytes));
+        Ok(())
+    })
+}
+
+/// Returns the next random u64 value from the ChaCha20Rng
+///
+/// /// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_next_long(
+    chacha20_rng: FfiMutPtr<Mutex<McChaCha20Rng>>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> u64 {
+    ffi_boundary_with_error(out_error, || Ok(chacha20_rng.lock()?.next_u64()))
+}
+
+/// frees the ChaCha20Rng
+///
+/// # Preconditions
+///
+/// * The ChaCha20Rng is no longer in use
+///
+/// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_free(
+    chacha20_rng: FfiOptOwnedPtr<Mutex<McChaCha20Rng>>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> bool {
+    ffi_boundary_with_error(out_error, || {
+        let _ = chacha20_rng;
+        Ok(())
+    })
```

Comment:
```suggestion
) {
    ffi_boundary(|| {
        let _ = chacha20_rng;
    })
```

Reply:
Updating as above - no need for the Ok(()) w/out error param (existing transaction.rs also does not rely on error param for its free call)